core/src/main/java/com/microsoft/applicationinsights/internal/shutdown/SDKShutdownActivity.java

```diff
@@ -126,23 +134,29 @@ private void stopInternalLogger() {
          * Make sure no exception is thrown!
          */
         private void stopChannels() {
+            for (TelemetryChannel channel : channels.keySet()) {
+                stopChannel(channel);
+            }
             for (ChannelFetcher fetcher : fetchers) {
+                stopChannel(fetcher.fetch());
+            }
+        }
+
+        private void stopChannel(TelemetryChannel channelToStop) {
+            try {
+                if (channelToStop != null) {
+                    channelToStop.stop(getPerThreadTimeout(), getPerThreadTimeUnit());
+                }
+            } catch (ThreadDeath td) {
+                throw td;
+            } catch (Throwable t) {
                 try {
-                    TelemetryChannel channelToStop = fetcher.fetch();
-                    if (channelToStop != null) {
-                        channelToStop.stop(getPerThreadTimeout(), getPerThreadTimeUnit());
-                    }
+                    InternalLogger.INSTANCE.error("Failed to stop channel: '%s'", t.toString());
+                    InternalLogger.INSTANCE.trace("Stack trace generated is %s", ExceptionUtils.getStackTrace(t));
```

Comment:
Since we're changing things, could you please consolidate these :) Also, I think this can be a warning. It's just to hint that there may be some threads hanging around.

Reply:
👍 Updated